app/models/criterion.rb

```diff
@@ -159,4 +162,47 @@ def self.update_assigned_groups_counts(assignment, criterion_ids_by_type = nil)
     #{"AND id IN (#{checkbox_criterion_ids_str})" unless checkbox_criterion_ids_str.empty?}
     UPDATE_SQL
   end
+
+  def replace_marks
+    mark_objects = []
+    # results with specific assignment
+    results = Result
+                .joins(submission: :grouping)
+                .where(groupings: {assignment_id: self.assignment_id})
+    if self.ta_visible_changed?
+      if self.ta_visible # The criterion has changed from not visible to visible
+        results.each do |r|
+          unless r.is_a_review? # filter results that are not peer reviews
+            mark_objects << Mark.new(result_id: r.id) # create mark object for TA review result
```

Comment:
This is not associating the mark object with the criterion (`self`). Use the appropriate method (see http://guides.rubyonrails.org/association_basics.html#has-many-association-reference) on `self.marks.____` to create the mark.